config/initializers/rack_attack.rb

```diff
@@ -1,124 +1,165 @@
   end
-  send(tracker, 'requests per unauthenticated user to history pdf', limit: 25, period: 10.seconds) do |request|
-    if tracking_enabled?(request)
-      if !warden_user_present?(request) && history_pdf_path(request)
-        request.ip
-      end
+
+  # is the source ip on a local or private network?
+  def trusted_proxy?
+    ActionDispatch::RemoteIp::TRUSTED_PROXIES.any? do |range|
+      range.include?(request.ip)
     end
   end
-  send(tracker, 'requests per unauthenticated user to assets', limit: 200, period: 10.seconds) do |request|
-    if tracking_enabled?(request)
-      if !warden_user_present?(request) && asset_paths(request)
-        request.ip
-      end
+end
+
+class Rack::Attack::Request
+  include RackAttackRequestHelpers
+end
+
+# rubocop:disable Style/IfUnlessModifier
+Rack::Attack.tap do |config|
+  # Throttling configuration
+  # * Multiple throttles can match the same request
+  # * Names must be unique or they will be silently overwritten
+
+  brute_force_options = {
+    limit: 20,
+    period: 3.minutes,
+  }
+
+  # Goal: prevent brute-force enumeration of credentials
+  config.throttle('per-ip limit on auth requests', **brute_force_options) do |request|
+    if request.tracking_enabled? && request.anonymous?
+      request.request_ip if request.hmis_authentication_attempt? || request.warehouse_authentication_attempt?
     end
   end
-  send(tracker, 'requests per logged-in user per ip', limit: 150, period: 5.seconds) do |request|
-    if tracking_enabled?(request)
-      if warden_user_present?(request) && !(rapid_paths(request))
-        request.ip
-      end
+
+  # Goal: prevent brute-force enumeration of password reset tokens
+  config.throttle('per-ip limit on password resets', **brute_force_options) do |request|
+    if request.tracking_enabled? && request.anonymous? && request.password_reset_attempt?
+      request.request_ip
     end
   end
-  send(tracker, 'requests per logged-in user per ip special', limit: 250, period: 5.seconds) do |request|
-    if tracking_enabled?(request)
-      if warden_user_present?(request) && (rapid_paths(request))
-        request.ip
-      end
+
+  # Goal: prevent brute-force enumeration of okta redirects
+  config.throttle('per-ip limit on okta redirects', **brute_force_options) do |request|
+    if request.tracking_enabled? && request.anonymous? && request.okta_callback?
+      request.request_ip
     end
   end
-  send(tracker, 'logins per account', limit: 10, period: 180.seconds) do |request|
-    if tracking_enabled?(request)
-      if sign_in_path(request) && user_email_present?(request)
-        request.params['user']['email']
-      end
+
+  # Goal: TBD
+  config.throttle(
+    ' per-ip limit on unauthenticated requests for history pdf',
+    limit: 25,
+    period: 10.seconds,
+  ) do |request|
+    if request.tracking_enabled? && request.anonymous? && request.history_pdf_path?
+      request_ip(request)
```

Comment:
why is this `request_ip(request)` but on line 147 we use `request.request_ip`?

Reply:
oops that's a bug and also points out I need to add a test